src/traits/trait-objects.md

```diff
@@ -0,0 +1,87 @@
+# Trait Objects
+
+We've seen how a function can take arguments which implement a trait:
+
+```rust
+trait Pet {
+    fn name(&self) -> String;
+}
+
+struct Dog {
+    name: String,
+}
+
+struct Cat;
+
+impl Pet for Dog {
+    fn name(&self) -> String {
+        self.name.clone()
+    }
+}
+
+impl Pet for Cat {
+    fn name(&self) -> String {
+        String::from("The cat") // No name, cats won't respond to it anyway.
+    }
+}
+
+fn greet(pet: &impl Pet) {
+    println!("Who's a cutie? {} is!", pet.name());
+}
+
+fn main() {
+    let pets: Vec<Box<dyn Pet>> = vec![
+        Box::new(Dog { name: String::from("Fido") }),
+        Box::new(Cat),
+    ];
+    for pet in pets {
+        println!("Hello {}!", pet.name());
+    }
+}
+```
+
+
+Memory layout after allocating `pets`:
+
+```bob
+ Stack                             Heap
+.- - - - - - - - - - - - - -.     .- - - - - - - - - - - - - - - - - - - - - - - - - -.
+:                           :     :                                                   :
+:    pets                   :     :                                                   :
+:   +-----------+-------+   :     :   +-----+-----+                                   :
+:   | ptr       |   o---+---+-----+-->| o o | o o |                                   :
+:   | len       |     2 |   :     :   +-|-|-+-|-|-+                                   :
+:   | capacity  |     2 |   :     :     | |   | |   +----+----+---+----+              :
+:   +-----------+-------+   :     :     | |   | '-->| F  | i  | d | o  |              :
+:                           :     :     | |   |     +----+----+---+----+              :
```

Comment:
This makes it look like it points to a 4-byte array. It points to a struct that in fact contains a 3-word String object.  maybe just

```
    +--------------+
--> | name: "Fido" |
    +--------------+
```

Reply:
Good point. Fixed.